### DISCUSSION TITLE
Find default presets in parent groups in GitLab

### DISCUSSION THREAD

**Discussion by valenvb** [renovatebot/renovate#25858] [Suggest an Idea]
### Tell us more.

The [organization level presets](https://docs.renovatebot.com/config-presets/#organization-level-presets) onboarding feature is awesome, but it would be great to take advantage of GitLab's ability to have nested groups of projects. For example, if CI container images live at `/my-org/my-team/container-images/node`, it would be ideal to allow having a default preset at the team level (since other teams in the org may want different configs) or even at the `container-images` level to have a single config specific to container projects.

I've put together a [proof of concept for this](https://github.com/renovatebot/renovate/compare/main...valenvb:renovate:feat/parent-group-presets), which seems to work well in GitLab. 

Basically, walking up the repo path hierarchy, choosing the first default config it comes across (i.e. the config closest to the current repo). It should also work for other platforms that support nested groups in this way, though I'm not sure if any others do.

**Comment by rarkins** [comment]
First question: before your changes, if the repo is `/my-org/my-team/container-images/node` then does Renovate look in `/my-org/my-team/container-images/renovate-config` and `/my-org/my-team/container-images/.gitlab` or in `/my-org/renovate-config` and `/my-org/.gitlab`? 

In the PoC it didn't appear that your loop was existing at the first config it finds, but I'm glad to see that are agree on that point.

The next point to discuss is the repo naming. Today Renovate supports `renovate-config` as well as `.platform`. We perhaps don't need to support both of them for this new nested default preset hunting. I think of the latter as mostly a github thing, and in that case there's no nested orgs anyway. What if we were to support only `renovate-config` hunting?

**Reply by valenvb** [reply]
> First question: before your changes, if the repo is /my-org/my-team/container-images/node then does Renovate look in /my-org/my-team/container-images/renovate-config and /my-org/my-team/container-images/.gitlab or in /my-org/renovate-config and /my-org/.gitlab?

It would look in `/my-org/renovate-config` and `/my-org/.gitlab`

> We perhaps don't need to support both of them for this new nested default preset hunting. I think of the latter as mostly a github thing, and in that case there's no nested orgs anyway. What if we were to support only renovate-config hunting?

I think supporting only `renovate-config` would be fine; you do occasionally see `.gitlab` projects containing things like group issue templates (though the name isn't a requirement), but they've also added the `gitlab-profile` repo for group readmes. Probably fine to support just a renovate-specific project until there's more cohesion around one or the other of those names. 

I also did a look through the rest of the supported platforms, and I think GitLab is the only one that does support nested groups, so it's probably reasonable to have a distinct codepath `if (GlobalConfig.get('platform') === 'gitlab')` and leave the current logic alone otherwise?

**Reply by rarkins** [reply]
So here's how I'd like it to work:
- It checks for a `renovate-config` repo from most specific (project) to least specific (top level group)
- If not found, it checks for `.platform` within top level group

This essentially retains backwards compatibility with today. I'm not sure you need to do any gitlab-specific logic because if the other platforms don't have hierarchy then the logic will remain the same

**Reply by valenvb** [reply]
That makes sense to me! I'm happy to adjust my initial implementation and open a PR to make these changes; should this discussion be turned into an issue that a PR can close first (I'm not particularly familiar with Github's discussions)?

**Comment by rarkins** [comment]
Converted to issue: #25885
